Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift

```
@@ -0,0 +1,274 @@
+    let stateMachine: StateMachine
+
+    // TODO: wrapped values are non-sendable
+    private let _client: NIOLockedValueBox<KafkaClient?>
+    private let _sequenceSource: NIOLockedValueBox<Producer.Source?>
+
+    /// The ``KafkaClient`` used for doing the actual polling.
+    var client: KafkaClient? {
+        get {
+            self._client.withLockedValue {
+                return $0
+            }
+        }
+        set {
+            self._client.withLockedValue {
+                $0 = newValue
+            }
+        }
+    }
+
+    /// The ``NIOAsyncSequenceProducer.Source`` used for yielding the messages to the ``NIOAsyncSequenceProducer``.
+    var sequenceSource: Producer.Source? {
+        get {
+            self._sequenceSource.withLockedValue {
+                return $0
+            }
+        }
+        set {
+            self._sequenceSource.withLockedValue {
+                $0 = newValue
+            }
+        }
+    }
+
+    private let logger: Logger
+
+    /// Initializes the ``KafkaBackPressurePollingSystem``.
+    ///
+    /// - Parameter logger: The logger to be used for logging.
+    init(logger: Logger) {
+        self.stateMachine = StateMachine()
+        self.logger = logger
+
+        self._client = NIOLockedValueBox(nil)
+        self._sequenceSource = NIOLockedValueBox(nil)
+    }
+
+    /// Runs the poll loop with the specified poll interval.
+    ///
+    /// - Parameter pollIntervalNanos: The poll interval in nanoseconds.
+    /// - Returns: An awaitable task representing the execution of the poll loop.
+    func run(pollIntervalNanos: UInt64) async {
+        actionLoop: while true {
+            let action = self.stateMachine.nextPollLoopAction()
+
+            switch action {
+            case .poll:
+                self.client?.withKafkaHandlePointer { handle in
+                    rd_kafka_poll(handle, 0)
+                }
+                try? await Task.sleep(nanoseconds: pollIntervalNanos)
```

We shouldn't silence the error here. If our task got cancelled we should exit out of the run loop. Importantly though we probably want to finish the source then.